proposals/0000-hypercore-header.md

```diff
@@ -0,0 +1,109 @@
+
+Title: **DEP-0000: Hypercore Header**
+
+Short Name: `0000-hypercore-header`
+
+Type: Standard
+
+Status: Undefined (as of 2018-06-26)
+
+Github PR: (add HTTPS link here after PR is opened)
+
+Authors: [Paul Frazee](https://github.com/pfrazee), [Mathias Buus](https://github.com/mafintosh)
+
+
+# Summary
+[summary]: #summary
+
+The Dat protocol supports a variety of data structures on top of the hypercore append-only logs, including key-value stores (hyperdb) and file archives (hyperdrive). In order for a program to read the data structure within a hypercore, it must know which structure it is reading. This DEP specifies a "header" entry which can be placed at the first entry of a hypercore in order to provide identifying information. This header also supports custom data for the data-structure, and may be expanded in future DEPs to provide additional standard data.
+
+
+# Motivation
+[motivation]: #motivation
+
+At time of writing, Dat's standard data structures are being expanded upon (hyperdb) while existing data structures are being updated with breaking changes (hyperdrive). In order for these changes to be deployed smoothly to the network, it's important that the following needs be met:
+
+ 1. Programs should correctly identify the data structure type and version within a hypercore. This will enable the program to use the correct handling code (eg hyperdb, hyperdrive-v1, hyperdrive-v2).
+ 2. Programs should abort processing a hypercore if they lack support for the type or version of the data structure.
+
+
+# Usage Documentation
+[usage-documentation]: #usage-documentation
+
+The "header" is the first entry in a hypercore. It includes a `type` and an optional `extension` which can contain data-structure-specific data.
+
+A program that is reading a hypercore will examine the `type` to determine how to process the hypercore.
+
+```js
+function loadCorrectStructure (hypercore, cb) {
+  readHeader(hypercore, (err, header) => {
+    if (err) return cb(err)
+
+    if (!header) {
+      // no header present - treat as a hypercore
+      return cb(null, hypercore)
+    }
+
+    switch (header.type) {
+      case 'hyperdrive':
+        return cb(null, createHyperdriveV1(hypercore, header))
+      case 'hyperdrive-v2':
+        return cb(null, createHyperdriveV2(hypercore, header))
+      case 'hyperdb':
+        return cb(null, createHyperdbV1(hypercore, header))
+      case '':
+        return cb(null, hypercore) // no structure, treat as a hypercore
+      // ...
+      default:
+        return cb(new Error('Unsupported: ' + header.type))
+    }
+  })
+}
+```
+
+The `type` string should be unique to the data structure (refer to existing DEPs to avoid conflicts). When a breaking change is made to the data-structure, it should be given a new `type` string. For example, the `"hyperdrive"` type string will be updated to `"hyperdrive-v2"` for its second version.
```

Comment:
Does this imply that the `-` character is special or reserved? Should successor versions use the `-vNNN` syntax?
To clarify my bikeshedding, I think encoding the version as an unstructured string is a reasonable choice, but I think a default delimiter between "schema" and "version/variant" should be chosen. 

Some common aesthetically pleasing separators would be `-`, `/`, `.`, `_`. How do you read `one-two_three-four`? As two identifiers (`one-two` followed by `three-four`), or three (`one`, `two_three`, `four`)?

To propose something concrete, lets say dash (`-`) is the separator, and identifiers should have dashes in them. Let's also say that the version is optional but recommended ("hyperdrive-v0"), and that the `x-` or `X-` prefixes are interpreted as "experimental" (aka, no DEP exists yet).

Reply:
I'm okay with some conventions but I'm not sure I want to standardize a syntax. Just not sure it's worth it. Reason being:

> Let's also say that the version is optional but recommended ("hyperdrive-v0"), and that the x- or X- prefixes are interpreted as "experimental" (aka, no DEP exists yet)

This is totally reasonable, but you then need to interpret `"hyperdrive-v0" == "hyperdrive"`, and I just don't know if we need that level of detail for this.